Create dependency-management.md (#3403)
dotnet/source-build: Documentation/dependency-management.md

***TODO: Add section about enabling dependency that's already available as source but not yet built in source build. ***

## I want to introduce a dependency on Foo. What do I do?

***TODO: Section about trying to avoiding new dependency when possible (i.e. by upgrading to a newer version or picking another library that's already available in the sources)***

## I really want to introduce dependency on Foo. What do I do?

The table below describes available methods of introducing dependencies in the product.
1. Evaluate each row to determine whether the method applies to your dependency.
2. For each applicable method, note which build configuration(s) it applies to.
3. After completing the evaluations, you should have at least one method that applies to source build, and one method that applies to Microsoft's propriety build. If you do not, please file an issue in the source-build repository.
4. Choose one method for each configuration. Using a single method is better.
5. Implement according to the guide below the table.

*Note: This list is not exhaustive and may not cover all cases. If you suspect your dependency is special, please contact **TODO***.

> | **Dependency Source** | **Description** | **Requirements** | **Configurations** |

Reviewer: Good idea.

| --------------------- | ----------------| ---------------- | --------------------------- |

> | [Source Build Externals (SBE)](https://github.com/dotnet/source-build-externals) | The source-build-externals repo contains the source of external .NET dependencies. The repository wraps and builds them according to source-build requirements. | <ul><li>Repository producing dependency is not on arcade</li><li>Dependency should not update often</li><li>NuGet package</li></ul> | Source-Build |

Reviewer 2: ```suggestion
| [Source Build Externals (SBE)](https://github.com/dotnet/source-build-externals) | The source-build-externals repo contains the source of external .NET dependencies. The repository wraps and builds them according to source-build requirements. | <ul><li>Repository producing dependency is not on arcade</li><li>Dependency should not update often</li><li>Dependency outputs one or more NuGet packages</li></ul> | Source-Build |
```

> | [Source Build Reference Packages (SBRP)](https://github.com/dotnet/source-build-reference-packages) | The source-build-reference packages repo contains source for the surface area of released .NET packages. | <ul><li>Dependency is contract-only. Not executed</li><li>Should not update often</li><li>Dependency is stable and released</li><li>Not already included in source-build-externals.</li><li>Package should be produced by .NET.</li><li>NuGet package.</li></ul> | Source-Build |

Reviewer 2: ```suggestion
| [Source Build Reference Packages (SBRP)](https://github.com/dotnet/source-build-reference-packages) | The source-build-reference packages repo contains source for the surface area of previously released .NET packages. | <ul><li>Dependency is contract-only. Not executed</li><li>Should not update often</li><li>Dependency is stable and released</li><li>Not already included in source-build-externals.</li><li>Package should be produced by .NET.</li><li>NuGet package.</li></ul> | Source-Build |
```

> | Previously source-built packages | As input to source-build, a set of previously source built packages are supplied, generated from the previous release or Microsoft's own release in certain bootstrapping scenarios. The versions of these inputs flow into the build, overriding elements in eng/Versions.props. | <ul><li>Input must be part of the previously source built product.</li><li>Input package must provide a compatible surface area.</li><li>Contents of package may not be redistributed into a shipping product asset.</li></ul> | Source-Build |

Reviewer 2: ```suggestion
| Previously source-built packages | As input to source-build, a set of previously source built packages are supplied, generated from the previous release or Microsoft's own release in certain bootstrapping scenarios. The versions of these inputs flow into the build, overriding elements in eng/Versions.props. | <ul><li>Input must be part of the previously source built product.</li><li>Input package must provide a compatible surface area.</li><li>Contents of package may not be redistributed into a shipping product asset (AKA [poisoning](leak-detection.md)).</li></ul> | Source-Build |
```
